Radiation rework and subsystem

ninjanomnom opened the pull request:
:cl: ninjanomnom
refactor: Radiation has been completely overhauled.
add: A new radiation subsystem and spreading mechanics.
add: Walls and other dense objects insulate you from radiation.
add: Geiger counters now store the last burst of radiation so you can view it at your leisure or show it to someone. Examine it.
add: Geiger counters can check mobs for contaminated objects. Scan yourself before you leave to make sure you aren't carrying dangerous radioactive items.
soundadd: Geiger counters have realistic sounds and the radiation pulse spam in chat has been replaced.
balance: Radiation is more deadly and causes burns at high intensities, WEAR YOUR RADSUITS.
balance: However residue radiation is far slower acting and kills you with toxin.
balance: Engineering holosigns have a light amount of radiation insulation.
balance: Rad collectors are nerfed. No more supercharging with pressurized plasma. No more collector spam either.
balance: Engine output is a lot more stable as a result of collector changes.
balance: Monkeys need more rads and take more time to turn into gorillas.
tweak: Over 100% on the DNA computer means your subject is now taking damage.
/:cl:

The goal of this is to remove snowflake applications of radiation and make it act realistically. With this system radiation can be insulated and it also now obeys the inverse square law. Overall this is a buff to radiation.

TODO:
- [x] Write the basic systems
- [x] Add radiation insulation to various turfs and structures
- [x] Find and convert all snowflake usages of radiation damage
- [x] Finish up radiation contamination
- [x] Reduce resource usage (Might need to do more)
- [x] Fix runtimes
- [x] Remap supermatter chambers (Will be done when the code is approved)
- [x] Test and balance

Issues:
- [x] ~~The SM gas interactions don't work well with power scaling~~ Fixed with some tweaks
- [x] ~~Contaminated tiles are building up rads when they shouldn't be~~ Bad use of log()
- [x] ~~Figure out exactly why the range gets so high (more insulation maybe?)~~ See above
- [x] ~~The high end is too high and the low end too low for mobs to handle well~~ Rewrote mob rad handling
- [x] ~~Genetics radiation needs to be fixed~~ Rad damage is matched with the others
- [x] ~~Mobs aren't getting toxin damage~~ Bad return

DoctoreMad commented:
Separation of alpha, beta and gamma radiation when?

ninjanomnom commented:
Out of scope. This doesn't change what radiation does, just how it acts.

ninjanomnom committed:
radiation rework and subsystem

ninjanomnom committed:
moves a few things to initialize and adds radiation insulation

Jordie0608 commented:
Half-life memes when?

ninjanomnom commented:
I could extremely easily make radioactive tainting with a half-life :^)

ChangelingRain commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -27,6 +27,12 @@
+	. = ..()
+	initial_state = icon_state
+	air_update_turf(1)
+	AddComponent(/datum/component/rad_insulation, RAD_MEDIUM_INSULATION)
```
make the insulation level a var and only create the component if there's insulation

optimumtact replied:
that's just needless var bloat

optimumtact commented on code/game/atoms.dm:
```diff
@@ -486,6 +486,10 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 	SendSignal(COMSIG_ATOM_EMAG_ACT)
 	return
 
+/atom/proc/rad_act(strength)
```
this is retarded, just call SendSignal on the atom

ninjanomnom replied:
This proc already exists and is used down the inheritance chain

Cyberboss replied:
since you're rewriting radiation you can get rid of the proc

ninjanomnom replied:
Alright, time for a bunch of new components then

Cyberboss replied:
HA RADtarded i just got that

Cyberboss commented on code/datums/components/rad_insulation.dm:
```diff
@@ -0,0 +1,6 @@
+/datum/component/rad_insulation //yes, this really is just a component to add a var
+	var/amount = 5
+
+/datum/component/rad_insulation/Initialize(insulation_amount)
```
set the default as the proc var default instead of on the datum

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -56,14 +55,20 @@
+		for(var/k in 1 to place.contents.len)
+			var/atom/thing = place.contents[k]
+			insulation = thing.GetComponent(/datum/component/rad_insulation)
+			//TODO: recursively loop through contents
```
Use place.GetAllContents

Cyberboss commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -151,18 +157,31 @@
+/obj/structure/mineral_door/uranium/Initialize()
+	. = ..()
+	var/datum/component/kill = GetComponent(/datum/component/rad_insulation)
+	kill.Destroy()
```
wat

Cyberboss replied:
sir, you know qdel exists right?

ninjanomnom replied:
...right, yes. I was rushing this part and was going off of component documentation looking for something like RemoveComponent() and it said destroy was public so I used it >.>

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -0,0 +1,79 @@
+
+	range_modifier = range_mod
+
+	SSradiation.processing += src
```
use START_PROCESSING

ninjanomnom committed:
adds a radiation contamination mechanic and makes requested changes

I'm terrified by the possibilities from this

ninjanomnom committed:
radiated objects give off light

and hopefuly contamination is balanced

banthebantz commented:
Fucking finally, this is very good

ninjanomnom committed:
fixes runtimes and an mc crash kek

removed the lighting part of contaminated objects
attempted some more balancing

ninjanomnom committed:
Collectors output power in process() gradualy

And some more balancing tweaks

Qustinnus commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,80 @@
+/datum/component/radioactive/Initialize(_strength=0, _half_life=RAD_HALF_LIFE, _can_glow=TRUE, _can_contaminate=TRUE)
+	. = ..()
+	strength = _strength
+	hl3_release_date = _half_life
```
wew lad

ninjanomnom committed:
Excludes a bunch of things from becoming radioactive

KorPhaeron commented:
Stale, tell me when it's done @ninjanomnom I was looking forward to this

ninjanomnom commented:
Repeating here what I said to kor in irc:

I'd prefer to not testmerge this sort of thing until the code has been mostly approved of by a maintainer because of the impact this could have on a round. And I'd also prefer to not do map changes until that happens. If I should go about this differently, let me know.

optimumtact commented:
can't review it without a test merge

ninjanomnom commented:
Fair enough I suppose, I'll work on the map changes tomorrow when I'm not on a caffeine life support at work.

Let's fuck up some poor engineer's day

ninjanomnom committed:
Ready for testmerge balancing

ninjanomnom committed:
Inverse square law was swapped

qbmax32 commented:
Live news from Bagil: It's randomly killing people and at least 15 people have dropped dead without warning. Fix plz XD

Jazaen commented:
Not only 15 people, but almost entire station. And after admins resurrect them, they walk a few steps and die again.

qbmax32 commented:
This just in: geigar counters recieving 124005 rads and everyone gets ded including the AI and borgs, Its outta control

Qustinnus commented:
guys protip: stand next to the SM and you'll be the safest. The further you're away, the deadlier it is. :^)

qbmax32 commented:
ill donate to FTL if you save us from this hell

Qustinnus commented:
alright it's fixed in the next update. Don't forget to donate.

ninjanomnom commented:
Get out of my pr floyd

Qustinnus commented:
never, I live here now.

Supermichael777 commented:
Literally instant death due to infectious radiation/10

ninjanomnom committed:
testmerge balancing

fixes the geiger counter
buffs collector power gen
nerfs sm radiation
slightly buffs rad insulation
nerfs rad damage (and its burn)
raised the minimum radiation

ninjanomnom commented:
changes for next testmerge update:

fixes the geiger counter
buffs collector power gen
nerfs sm radiation
slightly buffs rad insulation
nerfs rad damage (and its burn)
raised the minimum radiation

ninjanomnom committed:
disabling the radiation subsystem won't build up rad wave datums forever

WJohn commented:
>nuke ops land at the very north of the station as far as possible from the SM
>one of the ops INSTANTLY explodes from rads
>OH FUCK ABORT ABORT
>rest of the ops die seconds after returning to their base

good times

davethwave commented:
I wish I could have recorded that round, it was the only round that made me laugh so hard before.

GuyonBroadway commented:
I wanted the great uranium wall holocast to come back but this has gone from Holocaust to Imperial galactic ethnic cleansing program.

Mey-Ha-Zah commented:
I was a miner the shift the rad happened. I heard nobody talking over radio and went back to deposit my ores, wandered into the halls and saw three bodies and a red Geiger counter. My brain clicked and told me to run. I got about 8 tiles before the rad surge hit me and a died from like a million burn damage and was instantly husked.

10/10

ninjanomnom committed:
rewrites how mobs handle radiation

upgrades geiger counter functionality
and more balance tweaks

ninjanomnom committed:
cleans up stuff and removes debug message

ninjanomnom committed:
Slight contamination buff

ninjanomnom committed:
Major rad wave performance boost

Also improves rad insulation
Buffs contamination, again

ninjanomnom committed:
Fixes insulation runtime

More balance and performance tweaks

ninjanomnom committed:
fixes rad collectors not receiving power

Cyberboss commented:
How we doing here?

ninjanomnom committed:
The final balance commit

Fixes a major bug causing radiation to underperform
More geiger counter changes that will be changed more to add sounds

ninjanomnom commented:
It's been testmerged for a while which has been helpful for finding all the bugs and rad sources I missed. I still need to make showers clean contamination and fix genetic rad values. Then it's done. All the balancing is complete (At least enough for this pr)

I've got some planned stuff for radiation after this but that'll wait, this is a big enough pr as it is

ninjanomnom commented:
On second thought I'm leaving shower changes to the followup pr. Contamination goes away after a while anyway and showers are a mess of a code that need a rewrite; too big for this pr.

ninjanomnom committed:
Monkey business

ninjanomnom committed:
Geiger counter sounds

Cyberboss commented on code/__HELPERS/radiation.dm:
```diff
@@ -0,0 +1,35 @@
+		var/atom/thing = location.contents[i]
+		if(ignored_things[thing.type])
+			continue
+		get_rad_contents(thing, output)
```
can you make this iterative instead of recursive please

ninjanomnom replied:
The recursive method has been proven faster in #31027

Cyberboss commented on code/__HELPERS/radiation.dm:
```diff
@@ -0,0 +1,35 @@
+			continue
+		get_rad_contents(thing, output)
+
+/proc/radiation_pulse(turf/epicenter, intensity, range_modifier, log=0, can_contaminate=TRUE)
```
log = FALSE

Cyberboss commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -0,0 +1,26 @@
+SUBSYSTEM_DEF(radiation)
```
Use PROCESSING_SUBSYSTEM_DEF

Cyberboss commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -0,0 +1,26 @@
+	var/list/processing = list()
+	var/list/currentrun = list()
+
+/datum/controller/subsystem/radiation/fire(resumed=FALSE)
```
Then this entire proc isn't needed

Cyberboss commented on code/datums/components/rad_insulation.dm:
```diff
@@ -0,0 +1,10 @@
+	var/contamination_proof		// Can this object be contaminated?
+
+/datum/component/rad_insulation/Initialize(_amount=RAD_MEDIUM_INSULATION, _protects=TRUE, _contamination_proof=TRUE)
+	. = ..()
```
Initialize on components is an abstract proc so this isn't required

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+	var/can_contaminate
+
+/datum/component/radioactive/Initialize(_strength=0, _half_life=RAD_HALF_LIFE, _can_contaminate=TRUE)
+	. = ..()
```
ditto

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+			RegisterSignal(COMSIG_ITEM_ATTACK_OBJ, .proc/rad_attack)
+	else
+		warning("Something that wasn't an atom was given /datum/component/radioactive")
+		qdel(src)
```
CRASH instead of warning. I still haven't handled components deleting themselves in Initialize(). Not sure if I should

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+	return ..()
+
+/datum/component/radioactive/process()
+	if(QDELETED(parent))
```
Not possible

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+
+/datum/component/radioactive/InheritComponent(datum/component/C, i_am_original)
+	if(!i_am_original)
+		return ..()
```
..() undeeded

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+	if(!i_am_original)
+		return ..()
+	if(!hl3_release_date) // Permanently radioactive things don't get to grow stronger
+		return ..()
```
..() undeeded

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+		return ..()
+	var/datum/component/radioactive/other = C
+	strength = max(strength, other.strength)
+	return ..()
```
..() undeeded

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+/datum/component/radioactive/proc/rad_examine(mob/user, atom/thing)
+	var/out = ""
+	if(get_dist(parent, user) <= 1)
+		out += "The air around [parent] feels warm"
```
var/atom/P = parent to avoid datum var lookups

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+		if(RAD_AMOUNT_MEDIUM to RAD_AMOUNT_HIGH)
+			out += "[out ? " and it " : "[parent] "]seems to be glowing a bit."
+		if(RAD_AMOUNT_HIGH to INFINITY) //At this level the object can contaminate other objects
+			out += "[out ? " and it " : "[parent] "]hurts to look at."
```
replace these with it too

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+	return ..()
+
+/datum/component/radioactive/proc/rad_examine(mob/user, atom/thing)
+	var/out = ""
```
var/list/out = list()

Cyberboss commented on code/datums/components/radioactive.dm:
```diff
@@ -0,0 +1,77 @@
+			out += "[out ? " and it " : "[parent] "]hurts to look at."
+		else
+			out += "."
+	to_chat(user, out)
```
out.Join()

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -0,0 +1,97 @@
+
+	check_obstructions(atoms) // reduce our overall strength if there are radiation insulators
+
+	return TRUE
```
This retval isn't useful

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -0,0 +1,97 @@
+	var/list/atoms = list()
+	var/distance = steps
+
+	if(move_dir == NORTH || move_dir == SOUTH)
```
cache move_dir instead of reading it twice

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -0,0 +1,97 @@
+	if(move_dir == NORTH || move_dir == SOUTH)
+		distance-- //otherwise corners overlap
+
+	atoms += get_rad_contents(master_turf)
```
ditto for master_turf

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -0,0 +1,97 @@
+
+/datum/radiation_wave/proc/check_obstructions(list/atoms)
+	var/width = steps
+	if(move_dir == NORTH || move_dir == SOUTH)
```
see above

Cyberboss commented on code/datums/radiation_wave.dm:
```diff
@@ -0,0 +1,97 @@
+			continue
+		thing.rad_act(strength)
+
+		var/static/list/blacklisted = typecacheof(list(/turf))
```
I feel like this should be a GLOB typecache because I swear it's used more than just here

Cyberboss replied:
Also, just typecacheof(/turf)

ninjanomnom replied:
I'd prefer to leave this here because I'm pretty sure I'll need to add more entries

Cyberboss commented on code/game/atoms.dm:
```diff
@@ -485,6 +485,10 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 
+/atom/proc/rad_act(strength)
+	SendSignal(COMSIG_ATOM_RAD_ACT)
+	return TRUE
```
Can drop the return as it's not checked

Cyberboss commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -27,6 +27,12 @@
+/obj/structure/mineral_door/Initialize()
+	. = ..()
+	initial_state = icon_state
+	air_update_turf(1)
```
TRUE

Cyberboss commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -151,18 +157,30 @@
 
+/obj/structure/mineral_door/uranium/Initialize()
+	. = ..()
+	qdel(GetComponent(/datum/component/rad_insulation))
```
Instead of deleting new components like this, Add a common proc that you just override for /uranium

Cyberboss commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -226,6 +260,7 @@
 /obj/structure/mineral_door/paperframe/Initialize()
 	. = ..()
 	queue_smooth_neighbors(src)
+	qdel(GetComponent(/datum/component/rad_insulation))
```
ditto

Cyberboss commented on code/modules/power/supermatter/supermatter.dm:
```diff
@@ -584,7 +579,7 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_shard)
 
 	//Some poor sod got eaten, go ahead and irradiate people nearby.
-	radiation_pulse(get_turf(src), 4, 10, 500, 1)
+	radiation_pulse(get_turf(src), 500, 2, log=TRUE)
```
no need for named arg log

Cyberboss commented on code/modules/recycling/disposal-unit.dm:
```diff
@@ -40,6 +43,8 @@
 	//gas.volume = 1.05 * CELLSTANDARD
 	update_icon()
 
+	return INITIALIZE_HINT_LATELOAD //we need turfs to have air
```
The way this works is it asks SSatoms to call atom/LateInitialize() after it has Initialized all atoms. You didn't define one

ninjanomnom replied:
No it exists, I just moved this line from a duplicate Initialize

Cyberboss reviewed (changes requested):
I like it

ninjanomnom committed:
cleanup and move components to their own initialize

ninjanomnom committed:
Some code cleanup

And forgotten changes

Cyberboss commented on code/game/atoms.dm:
```diff
@@ -485,6 +492,10 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 
+/atom/proc/rad_act(strength)
+	SendSignal(COMSIG_ATOM_RAD_ACT)
+	return
```
trailing return

ninjanomnom committed:
Cleans up some trailing returns

ninjanomnom committed:
Mapping changes

GuyonBroadway commented:
Pretty rad.

davethwave commented:
congratulations

evsey9 commented:
REVERT THIS SHIT RIGHT FUCKING NOW
ITS BROKEN

Hathkar commented:
Currently extremely broken. A single pulse from genetics sends the rad levels to around 500%. Once you get any in your system, it's pretty much impossible to remove any. Revert or fix.

praisenarsie commented:
...this was merged 3 days ago, guys.

evsey9 commented:
Also, this breaks the singularity
Not enough power for the station.

davethwave commented:
perhaps make the singulo emit more rads or make rads spread with less loss in space to fix this?

ike709 commented:
There is already an open PR to fix genetics.

ike709 commented:
And issues open for the singularity.

msgerbs commented:
This is to this day the worst PR to ever grace TG.

optimumtact commented:
did you die to radiation?

GuyonBroadway commented:
I dunno, it's pretty rad.

msgerbs commented:
Damn right I did, except I didn't just die, I spent 20 minutes trying not to die, all because I tried to arrest somebody who had been in engineering. Making radiation some kind of contagious disease was and still is a horrible idea.

optimumtact commented:
he died :joy:

FancyCyclops commented:
This is how the fall began. How we went down the path of failure; the mark of the beginning of the end.

This PR and its subsequent merging was a mistake; it marks the beginning of a new era; an era of failure and headcoder griefing.

One can only pray for its excision from the code.
